src/Stratis.FederatedPeg.Features.FederationGateway/TargetChain/CrossChainTransferStore.cs

```diff
@@ -335,22 +441,74 @@ public async Task MergeTransactionSignaturesAsync(uint256 depositId, Transaction
         /// </summary>
         /// <param name="newTip">The new <see cref="ChainTip"/>.</param>
         /// <param name="blocks">The blocks used to update the store. Must be sorted by ascending height leading up to the new tip.</param>
-        private async Task PutAsync(HashHeightPair newTip, List<Block> blocks)
+        private void Put(HashHeightPair newTip, List<Block> blocks)
         {
-            Guard.NotNull(newTip, nameof(newTip));
-            Guard.NotNull(blocks, nameof(blocks));
             Guard.Assert(blocks.Count == 0 || blocks[0].Header.HashPrevBlock == (this.TipHashAndHeight?.Hash ?? 0));
 
             this.logger.LogTrace("()");
 
-            using (DBreeze.Transactions.Transaction transaction = this.DBreeze.GetTransaction())
+            using (DBreeze.Transactions.Transaction dbreezeTransaction = this.DBreeze.GetTransaction())
             {
-                transaction.SynchronizeTables(transferTableName, commonTableName);
-                await this.OnInsertBlocksAsync(transaction, newTip.Height - blocks.Count + 1, blocks);
+                dbreezeTransaction.SynchronizeTables(transferTableName, commonTableName);
+
+                int blockHeight = newTip.Height - blocks.Count + 1;
```

Comment:
I don't understand this line.  Also outside this method (see line 656) there is code to add the available blocks.  Unless I am missing something, it looks like the code here is removing them to calculate the correct height.  Would it make more sense to compute the newTip.Height outside of this method ?

Reply:
Thanks.